tslib.es6.js

```diff
@@ -107,14 +107,20 @@ export function __exportStar(m, exports) {
 }
 
 export function __values(o) {
-    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
+    var m = typeof Symbol === "function" && o[Symbol.iterator];
     if (m) return m.call(o);
-    return {
-        next: function () {
-            if (o && i >= o.length) o = void 0;
-            return { value: o && o[i++], done: !o };
-        }
-    };
+    if (typeof o === "object") {
+        if (Array.isArray(o)) {
+            var i = 0;
+            return {
+                next: function () {
+                    if (o && i >= o.length) o = void 0;
+                    return { value: o && o[i++], done: !o };
+                }
+            };
+        } else if (typeof o.next === "function") return o;

```

Comment:
ES2015 semantics do not support `for..of` on Iterators, only on Iterables.

Reply:
as per your comment, we should error here, I feel that the typescript documentation (https://www.typescriptlang.org/docs/handbook/compiler-options.html) and next version shout out should point out the fact that a polyfill is required for symbol and symbol.iterator as support generator return types has been improved in the next release so may trip other developers up as they start to use this functionality.  this is only really the case on cough IE or something like duktape/esprino, as all other browsers / node have had them for quite some time.

following on from your comments something like this?

```ts

  var m = typeof Symbol === "function" && o[Symbol.iterator];
            if (m) return m.call(o);

            if (o && typeof o.length !== "number") {
                var i = 0;
                return {
                    next: function () {
                        if (o && i >= o.length) o = void 0;
                        return { value: o && o[i++], done: !o };
                    }
                };
            } else if (typeof o === "object" && o.next === "function") {
                throw new TypeError(typeof o + " is iterable but there is no implementation of Symbol / Symbol.iterator (point to some documentation to dev experiance?) ");
            }

            throw new TypeError(typeof o + " is not iterable");

```